NeoMathEngine/src/CPU/CpuMathEngineDnnRowwise.h

```diff
@@ -0,0 +1,132 @@
+/* Copyright © 2017-2023 ABBYY
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+--------------------------------------------------------------------------------------------------------------*/
+
+#pragma once
+
+#include <NeoMathEngine/BlobDesc.h>
+
+namespace NeoML {
+
+// CPU implementation of rowwise operation
+class IRowwiseCpuImpl {
+public:
+	virtual ~IRowwiseCpuImpl() = default;
+
+	virtual int MinInputRowCount() const = 0;
+
+	virtual CBlobDesc Reshape( const CBlobDesc& inputSize ) = 0;
+	virtual int InOperationBufferSize() const = 0;
+	virtual int OutputRowCount() const = 0;
+	virtual int OutputRowSize() const = 0;
+	virtual bool IsInPlace() const = 0;
+
+	// The result of single rowwise processing
+	struct CProcessingReport {
+		int OutputRowsCalculated; // number of output rows calculated during this call
+		int InputRowsMayBeRemoved; // number of input rows which are not needed anymore
+	};
+
+	virtual CProcessingReport Process( const float* input, int inputRowIndex, int inputRowsAvailable,
+		float* output, int outputRowIndex, int outputRowsAvailable, float* buffer ) const = 0;
+};
+
+// Interface for buffers for rowwise operations
+// First DataRowCount() rows in buffer are filled with actual data
+// Rest of the rows are considered empty
+class IRowwiseBuffer {
+public:
+	virtual ~IRowwiseBuffer() = default;
+
+	// Number of elements in row
+	virtual int RowSize() const = 0;
+
+	// Index of first data row in buffer
+	virtual int DataRowIndex() const = 0;
+	// Number of rows in buffer filled with data
+	virtual int DataRowCount() const = 0;
+	// Number of data rows ever appeared in this buffer
+	virtual int DataRowProcessed() const { return DataRowIndex() + DataRowCount(); }
+	// Pointer to the beginning of rows with data
+	virtual const float* DataRows() const = 0;
+
+	// Empty rows are the rows which go immediately after the data
+	// Their number depends on the size of memory allocated for this buffer
+	// Number of empty rows in buffer
+	virtual int EmptyRowCount() const = 0;
+	// Pointer after data rows
+	virtual float* EmptyRows() = 0;
```

Comment:
`const`?

Reply:
I'm not so sure if it's a good idea. The data under this pointer is meant to be changed.

You fill it with some rows and mark these new rows as data rows via `AddRows`.